[WebAuthn] Do not request user presence before U2F fallback

pascoej opened the pull request:
#### ddef090f6b4d8984da6b8081120d40bf6e1a277e
<pre>
[WebAuthn] Do not request user presence before U2F fallback
<a href="https://rdar.apple.com/159976632">rdar://159976632</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=300839">https://bugs.webkit.org/show_bug.cgi?id=300839</a>

Reviewed by Brent Fulgham.

When authenticating with a security key supporting both CTAP2 and U2F, WebKit would
unnecessarily require two user taps and perform inefficient credential checking.

After CTAP2 silent credential checks failed, WebKit sent a CTAP2 request with user
presence (first tap), received an error, then downgraded to U2F (second tap). This
patch skips the wasteful CTAP2 request and downgrades immediately when silent checks
indicate no credentials exist, reducing taps from 2 to 1.

The batching logic incorrectly required both maxCredentialIDLength and
maxCredentialCountInList from getInfo to enable batching. The fix checks
for maxCredentialCountInList alone, enabling proper batching if maxCredentialIDLength
is missing.

Added layout tests to exercise this behavior.

* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-hid.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-hid.https.html:
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-u2f.https.html:
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::makeCredential):
(WebKit::CtapAuthenticator::getAssertion):
(WebKit::CtapAuthenticator::canDowngradeToU2f const):
(WebKit::CtapAuthenticator::tryDowngrade):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.h:

Canonical link: <a href="https://commits.webkit.org/301723@main">https://commits.webkit.org/301723@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ced7855c3ccac005fb0a1fe17213ba50d562e9c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126802 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46443 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37417 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133750 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78418 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128673 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47069 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54979 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96489 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/64519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129750 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37663 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113419 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77009 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36540 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31598 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77144 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107486 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31902 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136330 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53475 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41158 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104999 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53971 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109779 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104703 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26712 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50208 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28546 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50935 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53404 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59206 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52660 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55994 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54408 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
```diff
@@ -583,13 +599,11 @@ bool CtapAuthenticator::tryRestartPin(const CtapDeviceResponseCode& error)
-bool CtapAuthenticator::tryDowngrade()
+bool CtapAuthenticator::canDowngradeToU2f() const
 {
-    CTAP_RELEASE_LOG("tryDowngrade");
```
Is it worth leaving in as `CTAP_RELEASE_LOG("canDowngradeToU2f")`?

brentfulgham reviewed (approved):
Looks good! I had a minor suggestion, but otherwise no notes!

pascoej committed:
[WebAuthn] Do not request user presence before U2F fallback
rdar://159976632
https://bugs.webkit.org/show_bug.cgi?id=300839

Reviewed by Brent Fulgham.

When authenticating with a security key supporting both CTAP2 and U2F, WebKit would
unnecessarily require two user taps and perform inefficient credential checking.

After CTAP2 silent credential checks failed, WebKit sent a CTAP2 request with user
presence (first tap), received an error, then downgraded to U2F (second tap). This
patch skips the wasteful CTAP2 request and downgrades immediately when silent checks
indicate no credentials exist, reducing taps from 2 to 1.

The batching logic incorrectly required both maxCredentialIDLength and
maxCredentialCountInList from getInfo to enable batching. The fix checks
for maxCredentialCountInList alone, enabling proper batching if maxCredentialIDLength
is missing.

Added layout tests to exercise this behavior.

* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-hid.https-expected.txt:
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-hid.https.html:
* LayoutTests/http/wpt/webauthn/public-key-credential-get-success-u2f.https.html:
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::makeCredential):
(WebKit::CtapAuthenticator::getAssertion):
(WebKit::CtapAuthenticator::canDowngradeToU2f const):
(WebKit::CtapAuthenticator::tryDowngrade):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.h:

Canonical link: https://commits.webkit.org/301723@main
